pkg/scrape/discovery/file/file.go

```diff
@@ -0,0 +1,418 @@
+		RefreshInterval: model.Duration(5 * time.Minute),
+	}
+)
+
+func init() {
+	discovery.RegisterConfig(&SDConfig{})
+	prometheus.MustRegister(fileSDScanDuration, fileSDReadErrorsCount, fileSDTimeStamp)
+}
+
+// SDConfig is the configuration for file based discovery.
+type SDConfig struct {
+	Files           []string       `yaml:"files"`
+	RefreshInterval model.Duration `yaml:"refresh_interval,omitempty"`
+}
+
+// Name returns the name of the Config.
+func (*SDConfig) Name() string { return "file" }
+
+// NewDiscoverer returns a Discoverer for the Config.
+func (c *SDConfig) NewDiscoverer(opts discovery.DiscovererOptions) (discovery.Discoverer, error) {
+	return NewDiscovery(c, opts.Logger), nil
+}
+
+// SetDirectory joins any relative file paths with dir.
+func (c *SDConfig) SetDirectory(dir string) {
+	for i, file := range c.Files {
+		c.Files[i] = config.JoinDir(dir, file)
+	}
+}
+
+// UnmarshalYAML implements the yaml.Unmarshaler interface.
+func (c *SDConfig) UnmarshalYAML(unmarshal func(interface{}) error) error {
+	*c = DefaultSDConfig
+	type plain SDConfig
+	err := unmarshal((*plain)(c))
+	if err != nil {
+		return err
+	}
+	if len(c.Files) == 0 {
+		return errors.New("file service discovery config must contain at least one path name")
+	}
+	for _, name := range c.Files {
+		if !patFileSDName.MatchString(name) {
+			return errors.Errorf("path name %q is not valid for file discovery", name)
+		}
+	}
+	return nil
+}
+
+const fileSDFilepathLabel = model.MetaLabelPrefix + "filepath"
+
+// TimestampCollector is a Custom Collector for Timestamps of the files.
+type TimestampCollector struct {
+	Description *prometheus.Desc
+	discoverers map[*Discovery]struct{}
+	lock        sync.RWMutex
+}
+
+// Describe method sends the description to the channel.
+func (t *TimestampCollector) Describe(ch chan<- *prometheus.Desc) {
+	ch <- t.Description
+}
+
+// Collect creates constant metrics for each file with last modified time of the file.
+func (t *TimestampCollector) Collect(ch chan<- prometheus.Metric) {
+	// New map to dedup filenames.
+	uniqueFiles := make(map[string]float64)
+	t.lock.RLock()
+	for fileSD := range t.discoverers {
+		fileSD.lock.RLock()
+		for filename, timestamp := range fileSD.timestamps {
+			uniqueFiles[filename] = timestamp
+		}
+		fileSD.lock.RUnlock()
+	}
+	t.lock.RUnlock()
+	for filename, timestamp := range uniqueFiles {
+		ch <- prometheus.MustNewConstMetric(
+			t.Description,
+			prometheus.GaugeValue,
+			timestamp,
+			filename,
+		)
+	}
+}
+
+func (t *TimestampCollector) addDiscoverer(disc *Discovery) {
+	t.lock.Lock()
+	t.discoverers[disc] = struct{}{}
+	t.lock.Unlock()
+}
+
+func (t *TimestampCollector) removeDiscoverer(disc *Discovery) {
+	t.lock.Lock()
+	delete(t.discoverers, disc)
+	t.lock.Unlock()
+}
+
+// NewTimestampCollector creates a TimestampCollector.
+func NewTimestampCollector() *TimestampCollector {
+	return &TimestampCollector{
+		Description: prometheus.NewDesc(
+			"prometheus_sd_file_mtime_seconds",
```

Comment:
Just to avoid confusion, metrics are prefixed with `pyroscope`:
```suggestion
			"pyroscope_sd_file_mtime_seconds",
```